Use the new Preact CreateGroupForm for editing

seanh opened the pull request:
Use the new Preact-based `<CreateGroupForm />` component also for
editing groups.

seanh commented on h/static/scripts/group-forms/components/CreateGroupForm.tsx:
```diff
@@ -119,11 +119,14 @@ function TextField({
-  const [name, setName] = useState('');
-  const [description, setDescription] = useState('');
+  const config = useMemo(() => readConfig(), []);
+  const mode = config.components.CreateEditGroupForm.mode;
+  const group = config.components.CreateEditGroupForm.group;
```
We can probably do without the `mode` string and the component could instead react to the presence or absence of existing group details in the `js-config`

robertknight commented:
One way that we present success notifications in the client for various actions is to use a toast. A simple implementation of that pattern here would look like:

```diff
diff --git a/h/static/scripts/group-forms/components/CreateGroupForm.tsx b/h/static/scripts/group-forms/components/CreateGroupForm.tsx
index d7013a3c2..d26d2ada8 100644
--- a/h/static/scripts/group-forms/components/CreateGroupForm.tsx
+++ b/h/static/scripts/group-forms/components/CreateGroupForm.tsx
@@ -1,6 +1,13 @@
 import { useEffect, useId, useMemo, useState } from 'preact/hooks';
 
-import { Button, Input, Spinner, Textarea } from '@hypothesis/frontend-shared';
+import {
+  Button,
+  Input,
+  Spinner,
+  Textarea,
+  ToastMessages,
+  useToastMessages,
+} from '@hypothesis/frontend-shared';
 import { readConfig } from '../config';
 import { callAPI, CreateUpdateGroupAPIResponse } from '../utils/api';
 import { setLocation } from '../utils/set-location';
@@ -127,6 +134,8 @@ export default function CreateGroupForm() {
   const [description, setDescription] = useState(group.description);
   const [errorMessage, setErrorMessage] = useState('');
   const [saving, setSaving] = useState(false);
+  const { toastMessages, appendToastMessage, dismissToastMessage } =
+    useToastMessages();
 
   useEffect(() => {
     const listener = (e: PageTransitionEvent) => {
@@ -184,6 +193,10 @@ export default function CreateGroupForm() {
           description,
         },
       )) as CreateUpdateGroupAPIResponse;
+      appendToastMessage({
+        type: 'success',
+        message: 'Changes saved',
+      });
     } catch (apiError) {
       setErrorMessage(apiError.message);
     }
@@ -199,7 +212,7 @@ export default function CreateGroupForm() {
   }
 
   return (
-    <div className="text-grey-6 text-sm/relaxed">
+    <div className="relative text-grey-6 text-sm/relaxed">
       <h1 className="mt-14 mb-8 text-grey-7 text-xl/none">
         {mode === 'create'
           ? 'Create a new private group'
@@ -260,6 +273,13 @@ export default function CreateGroupForm() {
           &nbsp;Required
         </div>
       </footer>
+
+      <div className="absolute z-2 top-0 right-0">
+        <ToastMessages
+          messages={toastMessages}
+          onMessageDismiss={dismissToastMessage}
+        />
+      </div>
     </div>
   );
 }
```

With this pattern it is possible to dismiss toasts eagerly by calling `dismissToastMessage`, although I don't think it is critical to the usability of the form.

robertknight commented on h/static/scripts/group-forms/config.ts:
```diff
@@ -8,6 +8,18 @@ export type ConfigObject = {
+    updateGroup: APIConfig;
+  };
+  components: {
+    CreateEditGroupForm: {
```
I'm not convinced that structuring the config around the frontend's component structure is the right way to go, as it couples them unnecessarily.

seanh replied:
I've changed this to:

```javascript
export type ConfigObject = {
  ...
  context: {
    group: {
      pubid: string;
      name: string;
      description: string;
      link: string;
    } | null;
  };
};
```

robertknight commented on h/static/scripts/group-forms/config.ts:
```diff
@@ -8,6 +8,18 @@ export type ConfigObject = {
+        name: string;
+        description: string;
+        link: string;
+      };
```
All the group fields are empty for a new group, so an alternative way to represent this would be to simply make the `group` field null when creating a new group and omit the `mode` in that case. Something like:

```diff
diff --git a/h/static/scripts/group-forms/components/CreateGroupForm.tsx b/h/static/scripts/group-forms/components/CreateGroupForm.tsx
index d7013a3c2..6bd3fbcdb 100644
--- a/h/static/scripts/group-forms/components/CreateGroupForm.tsx
+++ b/h/static/scripts/group-forms/components/CreateGroupForm.tsx
@@ -120,11 +120,10 @@ function TextField({
 
 export default function CreateGroupForm() {
   const config = useMemo(() => readConfig(), []);
-  const mode = config.components.CreateEditGroupForm.mode;
-  const group = config.components.CreateEditGroupForm.group;
+  const group = config.group;
 
-  const [name, setName] = useState(group.name);
-  const [description, setDescription] = useState(group.description);
+  const [name, setName] = useState(group?.name ?? '');
+  const [description, setDescription] = useState(group?.description ?? '');
   const [errorMessage, setErrorMessage] = useState('');
   const [saving, setSaving] = useState(false);
 
@@ -192,16 +191,16 @@ export default function CreateGroupForm() {
   };
 
   let onSubmit;
-  if (mode === 'create') {
-    onSubmit = createGroup;
-  } else {
+  if (group) {
     onSubmit = updateGroup;
+  } else {
+    onSubmit = createGroup;
   }
 
   return (
     <div className="text-grey-6 text-sm/relaxed">
       <h1 className="mt-14 mb-8 text-grey-7 text-xl/none">
-        {mode === 'create'
+        {!group
           ? 'Create a new private group'
           : 'Edit group details'}
       </h1>
@@ -245,14 +244,14 @@ export default function CreateGroupForm() {
             disabled={saving}
             data-testid="button"
           >
-            {mode === 'create' ? 'Create group' : 'Save changes'}
+            {!group ? 'Create group' : 'Save changes'}
           </Button>
         </div>
       </form>
 
       <footer className="mt-14 pt-4 border-t border-t-text-grey-6">
         <div className="flex">
-          {mode === 'edit' && (
+          {group && (
             <a href={group.link}>← Back to group overview page</a>
           )}
           <div className="grow" />
diff --git a/h/static/scripts/group-forms/config.ts b/h/static/scripts/group-forms/config.ts
index 34c43db0c..22230971c 100644
--- a/h/static/scripts/group-forms/config.ts
+++ b/h/static/scripts/group-forms/config.ts
@@ -10,17 +10,12 @@ export type ConfigObject = {
     createGroup: APIConfig;
     updateGroup: APIConfig;
   };
-  components: {
-    CreateEditGroupForm: {
-      mode: string;
-      group: {
-        pubid: string;
-        name: string;
-        description: string;
-        link: string;
-      };
-    };
-  };
+  group: {
+    pubid: string;
+    name: string;
+    description: string;
+    link: string;
+  } | null;
 };
 
 /** Return the frontend config from the page's <script class="js-config">. */
diff --git a/h/templates/groups/create_edit.html.jinja2 b/h/templates/groups/create_edit.html.jinja2
index 713868e90..cae4d5f55 100644
--- a/h/templates/groups/create_edit.html.jinja2
+++ b/h/templates/groups/create_edit.html.jinja2
@@ -15,17 +15,7 @@
           "url": "{{ request.route_url("api.group", id=group.pubid) }}"
         }
       },
-      "components": {
-        "CreateEditGroupForm": {
-          "mode": "{{ mode }}",
-          "group": {
-            "pubid": "{{ group.pubid }}",
-            "name": "{{ group.name }}",
-            "description": "{{ group.description }}",
-            "link": "{{ group.link }}"
-          }
-        }
-      }
+      "group": {{ group|tojson }}
     }
   </script>
 {% endblock %}
diff --git a/h/views/groups.py b/h/views/groups.py
index 73a449f94..8a810e8c4 100644
--- a/h/views/groups.py
+++ b/h/views/groups.py
@@ -20,13 +20,7 @@ class GroupCreateEditController:
         """Render the page for creating a new group."""
         return {
             "page_title": "Create a new private group",
-            "mode": "create",
-            "group": {
-                "pubid": "",
-                "name": "",
-                "description": "",
-                "link": "",
-            },
+            "group": None
         }
 
     @view_config(
@@ -38,7 +32,6 @@ class GroupCreateEditController:
 
         return {
             "page_title": "Edit group details",
-            "mode": "edit",
             "group": {
                 "pubid": group.pubid,
                 "name": group.name,
```

The upside is that this explicitly models the possibility that there is no group, so e.g. it's harder to accidentally write code which uses `pubid` in a context where it may be empty.

The downside of this approach is that if you wanted to provide default values for fields, those would need to be defined in the frontend, or provided via a separate field.

seanh replied:
:+1: Done, I've removed the `mode` field in favor of reacting to `group` being truthy or falsey

robertknight commented on h/templates/groups/create_edit.html.jinja2:
```diff
@@ -11,6 +9,21 @@
+            "name": "{{ group.name }}",
+            "description": "{{ group.description }}",
+            "link": "{{ group.link }}"
+          }
```
I got confused initially while looking at this code because the shape of the `group` object is defined in both `groups.py` and here.

Separately, Jinja is a crap environment to define JSON structures in. I would suggest in future reworking the code so that we build up the configuration as a JSON-serializable Python dict and then the template just does `{{ config | tojson }}`. The code that generates the dict is then easy to execute in a test as well.

robertknight reviewed (commented):
I had a look over this and did some testing. Functionally everything works as expected. I made some notes around providing feedback after the form is saved and also the configuration structure.

seanh commented:
This is ready for re-review @robertknight. Changes:

* Removed the `mode` js-config setting in favour of reacting to the presence or absence of an exisintg `group` in the js-config instead
* Removed the name of the component from the js-config
* Moved the js-config generation into Python, so the Jinja2 template now just does `{{ js_config|tojson }}`
* Added a confirmation check mark that appears after successfully saving changes
* Added tests


https://github.com/user-attachments/assets/a592d091-745e-4ba3-b086-6cb3bf45834d

robertknight commented on h/static/scripts/group-forms/components/CreateEditGroupForm.tsx:
```diff
@@ -163,17 +183,52 @@ export default function CreateGroupForm() {
+        },
+      )) as CreateUpdateGroupAPIResponse;
+    } catch (apiError) {
+      setSaving(false);
```
The `setSaving(false)` call could be moved into a `finally` block, since such blocks always run, even if the try or catch blocks returns ([docs](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Statements/try...catch#the_finally_block)).

seanh replied:
Done

robertknight commented on h/static/scripts/group-forms/components/test/CreateEditGroupForm-test.js:
```diff
@@ -163,16 +178,21 @@ describe('CreateGroupForm', () => {
-    const { wrapper } = createWrapper();
-
     await assertInLoadingState(wrapper, false);
+    assert.isFalse(wrapper.exists('[data-testid="check"]'));
```
I often find it helpful to extract even these small `wrapper.exists`-type expressions into a helper (eg. `checkIsVisible(wrapper)`) to make it easier to update the selector if for any reason it needs to change in future.

seanh replied:
Done

robertknight commented on h/static/scripts/group-forms/components/test/CreateEditGroupForm-test.js:
```diff
@@ -235,6 +257,118 @@ describe('CreateGroupForm', () => {
+
+    it('shows an error message if callAPI() throws an error', async () => {
+      const errorMessageFromCallAPI = 'Bad API call.';
+      fakeCallAPI.rejects({ message: errorMessageFromCallAPI });
```
```suggestion
      fakeCallAPI.rejects(new Error(errorMessageFromCallAPI ));
```

As a matter of habit, it is a good idea to make sure promises always reject with instances of `Error`. In various contexts you get more useful output if this is the case.

seanh replied:
Done

robertknight reviewed (approved):
LGTM, this all works nicely. I think we want to introduce a short animation to make the saving check a little more eye-catching when it appears. That can be left as an exercise for a future PR.
